File: gcc_phat.py

import numpy as np


def gcc_phat(sig, refsig, fs=1, max_tau=None, interp=16):
    '''
    This function computes the offset between the signal sig and the reference signal refsig
    using the Generalized Cross Correlation - Phase Transform (GCC-PHAT)method.
    '''
    
    # make sure the length for the FFT is larger or equal than len(sig) + len(refsig)
    n = sig.shape[0] + refsig.shape[0]

    # Generalized Cross Correlation Phase Transform
    SIG = np.fft.rfft(sig, n=n)
    REFSIG = np.fft.rfft(refsig, n=n)
    R = SIG * np.conj(REFSIG)

    cc = np.fft.irfft(R / np.abs(R), n=(interp * n))

    if not max_tau:
        max_tau = n / 2.0

    max_shift = int(interp * fs * max_tau)

    cc = np.concatenate((cc[-max_shift:], cc[:max_shift+1]))

    # find max cross correlation index
    shift = np.argmax(np.abs(cc))
    tau = (shift - max_shift) / float(interp * fs)
    
    return tau


def main():
    
    refsig = np.linspace(1, 10, 10)

    for i in range(0, 10):
        sig = np.concatenate((np.linspace(0, 0, i), refsig, np.linspace(0, 0, 10 - i)))
        print(gcc_phat(sig, refsig))

    
if __name__ == "__main__":
    main()

File: octave/gcc_phat.m
function [tau, cc] = gcc_phat(sig, refsig, fs, max_tau, interp)

n = length(sig) + length(refsig);
max_shift = floor(n / 2);

if nargin < 3
  fs = 1;
end

if nargin >= 4
  max_shift = min(floor(fs * max_tau), max_shift);
end

if nargin < 5
  interp = 1;
end

max_shift = max_shift * interp;

X1 = fft(sig, n);
X2 = fft(refsig, n);
R = X1 .* conj(X2);

cc = ifft(R ./ (abs(R)), n * interp);
N = length(cc);
cc = [cc((N - max_shift + 1):N)(:); cc(1:(max_shift + 1))(:)];
cc = abs(cc);

[max_cc, shift] = max(cc);
shift -= max_shift + 1;

tau = shift / (interp * fs);

end



